File: easy/0408_valid_word_abbreviation/solution.cpp
#include <string>

class Solution {
public:
  bool validWordAbbreviation(std::string word, std::string abbr) {
    const size_t wordLen = word.size();
    const size_t abbrLen = abbr.size();

    size_t wordI = 0;
    size_t abbrI = 0;

    while (wordI < wordLen) {
      if (word[wordI] == abbr[abbrI]) {
        ++wordI;
        ++abbrI;
      } else if (abbr[abbrI] >= '0' && abbr[abbrI] <= '9') {
        if (abbr[abbrI] == '0')
          return false;

        size_t jump = 0;

        while (abbrI < abbrLen && std::isdigit(abbr[abbrI]))
          jump = jump * 10 + (abbr[abbrI++] - '0');

        wordI += jump;
      } else {
        return false;
      }
    }

    return wordI == wordLen && abbrI == abbrLen;
  }
};

File: easy/0408_valid_word_abbreviation/Solution.java
class Solution {
  public boolean validWordAbbreviation(String word, String abbr) {
    final int wordLen = word.length();
    final int abbrLen = abbr.length();

    int wordI = 0;
    int abbrI = 0;

    while (wordI < wordLen) {
      if (word.charAt(wordI) == abbr.charAt(abbrI)) {
        ++wordI;
        ++abbrI;
      } else if (abbr.charAt(abbrI) >= '0' && abbr.charAt(abbrI) <= '9') {
        if (abbr.charAt(abbrI) == '0') {
          return false;
        }

        int jump = 0;

        while (abbrI < abbrLen && Character.isDigit(abbr.charAt(abbrI))) {
          jump = jump * 10 + abbr.charAt(abbrI) - '0';
          ++abbrI;
        }

        wordI += jump;
      } else {
        return false;
      }
    }

    return wordI == wordLen && abbrI == abbrLen;
  }
}
